### PR TITLE
Fix incorrect phase function blending in multi-component atmospheres

### PR DESCRIPTION
**Pull request by nollety**
# Description

Closes eradiate/eradiate-issues#155

# Checklist

- [x] The code follows the relevant coding guidelines
- [x] The code generates no new warnings
- [x] The code is appropriately documented
- [x] The code is tested to prove its function
- [x] The feature branch is rebased on the current state of the `main` branch
- [x] I give permission that the Eradiate project may redistribute my contributions under the terms of its license
- [x] :warning: update regression test: `tests/02_eradiate/03_regression/atmospheres/test_rpv_afgl1986_continental.py::test_rpv_afgl1986_continental_brfpp`

### CODE REVIEW

**Comment by nollety** [comment]
- [ ] update regression test: `tests/02_eradiate/03_regression/atmospheres/test_rpv_afgl1986_continental.py::test_rpv_afgl1986_continental_brfpp`

File: src/eradiate/scenes/phase/_blend.py [hunk 1]
```diff
@@ -15,6 +16,23 @@
 from ...units import unit_context_kernel as uck
 
 
+def _convert_weights(value: npt.ArrayLike) -> npt.NDArray[np.float64]:
```
Comment (leroyvn) [inline src/eradiate/scenes/phase/_blend.py:19]: For consistency with the `attrs` decorator naming (rule used elsewhere in the code base; the rest is a consequence of my comment on Numpy typing):
```suggestion
def _weights_converter(value: np.typing.ArrayLike) -> np.ndarray:
```

File: src/eradiate/scenes/phase/_blend.py [hunk 2]
```diff
@@ -4,6 +4,7 @@
 
 import attr
 import numpy as np
+import numpy.typing as npt
```
Comment (leroyvn) [inline src/eradiate/scenes/phase/_blend.py:7]:
```suggestion
```

File: src/eradiate/scenes/phase/_blend.py [hunk 3]
```diff
@@ -199,7 +211,7 @@ def kernel_dict(self, ctx: KernelDictContext) -> KernelDict:
         if self.weights.ndim == 2 and self.weights.shape[1] > 1:
             write_binary_grid3d(
                 filename=self.weight_file,
-                values=weights[np.newaxis, np.newaxis, ...],
+                values=weight[np.newaxis, np.newaxis, ...],
```
Comment (leroyvn) [inline src/eradiate/scenes/phase/_blend.py:214]: Let's take the opportunity to use a nicer syntax:
```suggestion
                values=np.reshape(weight, (1, 1, -1)),
```

File: src/eradiate/scenes/phase/_blend.py [hunk 4]
```diff
@@ -15,6 +16,23 @@
 from ...units import unit_context_kernel as uck
 
 
+def _convert_weights(value: npt.ArrayLike) -> npt.NDArray[np.float64]:
+    """
+    Convert weights.
+
```
Comment (leroyvn) [inline src/eradiate/scenes/phase/_blend.py:22]: Obvious given the function name.
```suggestion
```

File: src/eradiate/scenes/phase/_blend.py [hunk 5]
```diff
@@ -163,33 +181,27 @@ def _gridvolume_transform(self) -> "mitsuba.core.ScalarTransform4f":
 
     def kernel_dict(self, ctx: KernelDictContext) -> KernelDict:
         # Build kernel dict recursively
-        weights = self.weights[0, ...]
+
+        # Weight of all nested phase function but the first one
+        # (in the case where there are two nested phase functions,
+        # this is the weight of the second nested phase function):
+        weight = self.weights[1:, ...].sum(axis=0)
```
Comment (leroyvn) [inline src/eradiate/scenes/phase/_blend.py:188]:
```suggestion
        # Summed weights of all components except the first, used to 
        # set up the kernel dictionary of the phase function
        weight = self.weights[1:, ...].sum(axis=0)
```

File: src/eradiate/scenes/phase/_blend.py [hunk 6]
```diff
@@ -163,33 +181,27 @@ def _gridvolume_transform(self) -> "mitsuba.core.ScalarTransform4f":
...
-            marginal_weights = np.divide(
-                sub_weights,
-                sum_sub_weights,
-                where=sum_sub_weights != 0.0,
-                out=np.ones_like(sub_weights),
-            )
-
+            # All nested phase function but the first one are aggreated
+            # in a blend phase function, that is nested into the current
+            # blend phase function.
```
Comment (leroyvn) [inline src/eradiate/scenes/phase/_blend.py:197]: Be more concise, the code explains this already:
```suggestion
            # The phase function corresponding to remaining components
            # is built recursively
```

File: tests/02_eradiate/01_unit/scenes/phase/test_blend.py [hunk 7]
```diff
@@ -144,3 +186,65 @@ def test_blend_array(modes_all):
...
+            {"type": "isotropic"},
+        ],
+        weights=[weight1, weight2, weight3],
+        bbox=[[0, 0, 0], [1, 1, 1]],
+    )
+
+    # Kernel dict generation succeeds
+    ctx = KernelDictContext()
+    phase_dict = phase.kernel_dict(ctx)
+    assert phase_dict.load()
```
Comment (leroyvn) [inline tests/02_eradiate/01_unit/scenes/phase/test_blend.py:238]: I'd like to verify the type of the created object:
```suggestion
    assert phase_dict.load() is mi.PhaseFunction
```

File: tests/02_eradiate/01_unit/scenes/phase/test_blend.py [hunk 8]
```diff
@@ -86,38 +87,79 @@ def test_blend_bbox(mode_mono):
...
+        weights=[weight1, weight2],
     )
     kernel_dict = phase.kernel_dict(ctx)
     phase_dict = onedict_value(kernel_dict.data)
     assert phase_dict["type"] == "blendphase"
     assert phase_dict["phase1"] == {"type": "isotropic"}
     assert phase_dict["phase2"] == {"type": "rayleigh"}
-    assert phase_dict["weight"] == 0.3
+    assert phase_dict["weight"] == weight2
     assert kernel_dict.load()
```
Comment (leroyvn) [inline tests/02_eradiate/01_unit/scenes/phase/test_blend.py:117]:
```suggestion
    assert kernel_dict.load() is mi.PhaseFunction
```

File: tests/02_eradiate/01_unit/scenes/phase/test_blend.py [hunk 9]
```diff
@@ -86,38 +87,79 @@ def test_blend_bbox(mode_mono):
...
+            weight3,
+            weight2 + weight3,
+            where=weight2 + weight3 != 0.0,
+            out=np.zeros_like(weight3),
+        )
+    )
+    assert phase_dict["phase2"]["weight"] == nested_weight
     assert phase_dict["phase2"]["phase1"] == {"type": "rayleigh"}
     assert phase_dict["phase2"]["phase2"] == {"type": "hg", "g": 0.1}
     assert kernel_dict.load()
```
Comment (leroyvn) [inline tests/02_eradiate/01_unit/scenes/phase/test_blend.py:165]:
```suggestion
    assert kernel_dict.load() is mi.PhaseFunction
```

File: tests/02_eradiate/01_unit/scenes/phase/test_blend.py [hunk 10]
```diff
@@ -144,3 +186,65 @@ def test_blend_array(modes_all):
     # Kernel dict generation succeeds
     ctx = KernelDictContext()
     assert phase.kernel_dict(ctx).load()
```
Comment (leroyvn) [inline tests/02_eradiate/01_unit/scenes/phase/test_blend.py:188]:
```suggestion
    assert phase.kernel_dict(ctx).load() is mi.PhaseFunction
```

File: tests/02_eradiate/01_unit/scenes/phase/test_blend.py [hunk 11]
```diff
@@ -144,3 +186,65 @@ def test_blend_array(modes_all):
...
+            {"type": "rayleigh"},
+        ],
+        weights=[weight1, weight2],
+        bbox=[[0, 0, 0], [1, 1, 1]],
+    )
+
+    # Kernel dict generation succeeds
+    ctx = KernelDictContext()
+    phase_dict = phase.kernel_dict(ctx)
+    assert phase_dict.load()
```
Comment (leroyvn) [inline tests/02_eradiate/01_unit/scenes/phase/test_blend.py:210]:
```suggestion
    assert phase_dict.load() is mi.PhaseFunction
```

**Review by leroyvn — CHANGES_REQUESTED** [review]
Hi @ nollety, I think your diagnostic is correct. Hereafter are a few comments on your changes—overall, mostly style issues.

**Comment by nollety** [comment]
Hi @leroyvn, thanks for the feedback, I applied all your suggestions!

I'll update the failing regression test reference data set and after that, this pull request should be ready.

These assertions did not work:
```python
assert kernel_dict.load() is mi.PhaseFunction
```
so I changed them into:
```python
assert isinstance(kernel_dict.load(), mi.PhaseFunction)
```
or did you have something different in mind?

**Comment by nollety** [comment]
@leroyvn I have updated the reference dataset for the failing regresion test, all test pass; the pull request is ready :smiley:

**Comment by nollety** [comment]
I updated the changelog

**Comment by leroyvn** [comment]
> These assertions did not work:
> 
> ```python
> assert kernel_dict.load() is mi.PhaseFunction
> ```

My mistake, I reviewed this in a hurry and you did the right thing.

**Pull request merged**